sycl/source/detail/memory_manager.cpp

```diff
@@ -990,6 +991,164 @@ void MemoryManager::memset_2d_usm(void *DstMem, QueueImplPtr Queue,
       Height, DepEvents.size(), DepEvents.data(), OutEvent);
 }
 
+static void memcpyToDeviceGlobalUSM(QueueImplPtr Queue,
+                                    DeviceGlobalMapEntry *DeviceGlobalEntry,
+                                    size_t NumBytes, size_t Offset,
+                                    const void *Src,
+                                    std::vector<RT::PiEvent> DepEvents,
+                                    RT::PiEvent *OutEvent) {
+  // Zero-initialization is only needed if we do not overwrite the entire
+  // device_global.
+  bool NeedZeroInit =
+      Offset != 0 || NumBytes != DeviceGlobalEntry->MDeviceGlobalTSize;
```

Comment:
What would happen in this scenario:
```
// Non-initialized global state here.
memcpyToDevice(full global) // first operation
memcpyToDevice(part global) // << here?
```

I would imagine ZeroInit Event might need to be Last/FullInit event instead, but don't have a clear picture in my mind yet.

I also think that I'd prefer not having this optimization and introduce an explicit user-available property to skip zero-init when it matters.

Reply:
After an offline discussion I concede there are certain cases (especially possible through interop) where the operations could happen out-of-order, in which case the fill operation should happen even if the full memcpy was the first to be queued. As such it is now mandatory. Hopefully we can cover the common case with `consteval` in the future!